### PR TITLE
wrappers: symlink completion snippets when symlinking binaries

### PR DESCRIPTION
**Pull request by chipaca**
This also modifies complete.sh to support this, as well as some minor
changes to dirs and snap to accommodate.

Symlinking the snippets is itself simple enough that I didn't think it
was worth creating an ad-hoc helper, so it's just a few lines glommed
onto the end of the existing wrapper funcs (and tests).

> To be clear, I did write the helper, but between it being very few
> lines of code, it needing to return enough information to the
> caller to clean up behind it if something else went wrong, and
> especially as the other user of that helper would be the aliases
> and they don't have access to the AppInfo (which would be the
> obvious thing to pass to such helpers), I decided against it.

How aliases are going to play into completion is still an open
question on the implementation side, which I need to look into soon,
but this is not that work.

### CODE REVIEW

File: wrappers/binaries.go [hunk 1]
```diff
@@ -22,11 +22,15 @@ package wrappers
...
 import (
 	"os"
+	"path/filepath"
 
 	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/snap"
 )
 
+var completeSh = filepath.Join(dirs.SnapMountDir, "core", "current", "usr", "lib", "snapd", "complete.sh")
```
Comment (pedronis) [inline wrappers/binaries.go]: shouldn't this live in dirs too?

Comment (zyga) [inline wrappers/binaries.go]: Note that SnapMountDir is different on the inside and outside so be careful please.

Comment (chipaca) [inline wrappers/binaries.go]: about it living in dirs, maybe?
but yes this needs to be the "outside" one (which it is; the other one has Core in its name)

File: wrappers/binaries.go [hunk 2]
```diff
@@ -56,15 +58,34 @@ func AddSnapBinaries(s *snap.Info) (err error) {
...
+			return err
+		}
 	}
 
 	return nil
 }
 
 // RemoveSnapBinaries removes the wrapper binaries for the applications from the snap which aren't services from.
 func RemoveSnapBinaries(s *snap.Info) error {
+	noCompletion := !osutil.FileExists(dirs.CompletersDir) || !osutil.FileExists(dirs.CompleteSh)
```
Comment (pedronis) [inline wrappers/binaries.go]: shouldn't the remove case just check for CompletersDir ?

File: data/completion/complete.sh [hunk 3]
```diff
@@ -100,17 +101,28 @@ _complete_from_snap() {
...
-        _complete_from_snap "$1"
-        return $?
-    fi
-    # fallback to the old -D
-    _completion_loader "$1"
-}
+# this file can be sourced directly as e.g. /usr/lib/snapd/complete.sh, or via
+# a symlink from /usr/share/bash-completion/completions/. In the first case we
+# want to load the default loader; in the second, the specific one.
+if [[ "${BASH_SOURCE[0]}" =~ /complete\.sh$ ]]; then
```
Comment (pedronis) [inline data/completion/complete.sh]: what happens if somebody has  a command called "complete.sh" for their snap? I think it's not valid app command but it's a valid alias, so something to consider

File: wrappers/binaries.go (hunk 2 above)
Comment (chipaca) [inline wrappers/binaries.go]: actually, remove could just readlink

File: data/completion/complete.sh (hunk 3 above)
Comment (chipaca) [inline data/completion/complete.sh]: good point. Sounds like something I'd do :-D
I guess I'll have to put the full path in there.

Comment (chipaca) [inline data/completion/complete.sh]: the full path to complete.sh is Hard, so I switched the logic around and am checking for the snippet (those are always in `/usr/share/bash-completion/completions`, phew)

**Pull request merged**